### DISCUSSION TITLE
Including mongoose.h in a Windows C++ project breaks other includes and has many conflicts with winsock

### DISCUSSION THREAD

**Discussion by ThomasR95** [cesanta/mongoose#2985] [Ask questions, get answers] [answered]
Hi, I'm trying to add mongoose to an existing project.
Before I added mongoose, the project built without errors.
After including mongoose.h, I get many errors about redefinitions and missing declarations.
![image](https://github.com/user-attachments/assets/af5942c1-c786-4961-a304-01ea0913e98d)
...etc.
I've followed the guide to the best of my ability. I started by simply including mongoose.h and mongoose.c in my Visual Studio project, but when this returned errors, I went to the internet for advice and found the idea of building mongoose as a separate library and linking to it, to hopefully avoid such conflicts. Unfortunately that didn't make any difference.
Here are my includes in main.cpp:
```cpp
#include <string>
#include <iostream>
#include <random>
#include <deque>

#define IMGUI_DISABLE_WIN32_FUNCTIONS
#define IMGUI_ENABLE_FREETYPE
#include "imgui.h"
#include "imgui/misc/freetype/imgui_freetype.h"
#include "imgui-SFML.h"
#include "imgui_internal.h"

#include "defines.h"
#include <fstream>

#ifdef _WIN32
#include "spout2/SpoutSender.h"
#include "wtypes.h"
#include <Windows.h>
#include <fileapi.h>
#include <Dwmapi.h>
#include "CrashHandler.h"

#pragma comment (lib, "Dwmapi.lib")
#endif

#include "Config.h"
#include "xmlConfig.h"
#include "LayerManager.h"

// must be last?
#include "websocket.h"
```

and here is websocket.h:
```cpp
#pragma once
#include <iostream>

#include "mongoose.h"

void ev_handler(struct mg_connection* c, int ev, void* ev_data)
{
if (ev == MG_EV_HTTP_MSG)
{
	struct mg_http_message* hm = (struct mg_http_message*)ev_data;

	std::cout << hm->message.buf << std::endl;
}
}

class WebSocket
{

public:
void Start()
{
	mg_mgr_init(&eventManager);
	mg_http_listen(&eventManager, "http://0.0.0.0:8000", ev_handler, NULL);
}

void Poll()
{
	mg_mgr_poll(&eventManager, 1000);
}

~WebSocket()
{
	mg_mgr_free(&eventManager);
}

private:

struct mg_mgr eventManager;

};
```

Please could somebody help me get this building? Thanks in advance!

**Comment by scaprile** [comment]
TL;DR
Please be kind to us and clearly explain which warnings/errors you have without forcing us to try to read from a snapshot. Please correctly paste the warning text, and please don't send long code, we are busy writing Mongoose code.
Also, try to see why Visual Studio does not like that, because we test with minGW++ every night

**Reply by ThomasR95** [reply]
Thanks for the response.
> which warnings/errors you have

There are over 100 errors, which is why I didn't originally post them all to avoid overwhelming you.
Here are a few unique ones which dominate most of that list:
```1>D:\OneDrive\Rahituber_project\rahituber\RahiTuber\defines.h(542,13): error C2039: 'map': is not a member of 'std'
1>(compiling source file 'main.cpp')
1>    C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\include\locale(24,1):
1>    see declaration of 'std'

1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\shared\ws2def.h(638,28): error C2011: 'sockaddr_in': 'struct' type redefinition
1>(compiling source file 'main.cpp')
1>    C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winsock.h(309,8):
1>    see declaration of 'sockaddr_in'

1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\shared\ws2ipdef.h(81,24): error C2079: 'sockaddr_gen::AddressIn' uses undefined struct 'sockaddr_in'
1>(compiling source file 'main.cpp')

1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\ws2tcpip.h(982,27): error C2568: '=': unable to resolve function overload
1>(compiling source file 'main.cpp')
1>    C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\ws2tcpip.h(982,27):
1>    could be 'int getsockopt(SOCKET,int,int,char *,int *)'
```
- note that some things in std are seemingly getting redefined (or undefined?).
- sockaddr_in (for example, there are many other cases) is simultaneously defined too many times and also undefined

> don't send long code

I sent you all the code I thought would be necessary to demonstrate where I have included mongoose.h (in websocket.h, which just contains the basic snippets from your readme rearranged into a c++ class), and the reason I showed the includes in my main.cpp is to demonstrate that I have included mongoose as late as possible to try to avoid conflicts. I thought showing the other includes might also be useful in case there are any known conflicts you can see in that list.

> try to see why Visual Studio does not like that

I can successfully build mongoose.c separately using the following commands through the Visual Studio 2022 build tools (which I am doing to hopefully avoid the conflicts and redefinitions, but I think I've misunderstood something somewhere): 
```
cl /c /EHsc mongoose.c
lib mongoose.obj
```
So it doesn't seem that visual studio has a problem with mongoose, only that simply including mongoose in the source tree of my C++ project as stated in your guide doesn't seem to be the entire necessary setup in my case.

I'm not suggesting that mongoose is actually actively breaking this stuff, but just looking for help integrating it because somehow the steps in the guide didn't work for me.
Thanks!

**Reply by scaprile** [reply]
> There are over 100 errors,

Well, if you ever worked with C, you know that it is usually the first one you should see.

Mongoose is C, not C++, whatever you are doing when you include it triggers something that your compiler does not like.

**Reply by ThomasR95** [reply]
Clearly. What I'm hoping for is some insight on how including it might be upsetting the compiler. I've already sent what I'm doing to include it. 

The first error is that std::map undefined issue. Doesn't tell me much, as std::map is defined when I don't include mongoose. 

I'm aware that mongoose is C but your website says it works in C++, and doesn't provide any special instructions for integrating with C++. I've tried putting `#include "mongoose.h"` inside `extern "C"` but that didn't help.

**Reply by ThomasR95** [reply]
When I explicitly include <map> again to get rid of that error (not sure why it appeared when it used to be working), the new first error is this:
```
1>C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\shared\ws2def.h(240,25): error C2011: 'sockaddr': 'struct' type redefinition
1>(compiling source file 'main.cpp')
1>    C:\Program Files (x86)\Windows Kits\10\Include\10.0.22621.0\um\winsock.h(482,8):
1>    see declaration of 'sockaddr'
```
I'm not including either ws2def or winsock myself, so I assume that's mongoose's doing, but I am including Windows.h for some other platform-specific things. 

For anyone else who might run into this, here's the solution:

I found that adding #define WIN32_LEAN_AND_MEAN before my Windows.h inclusion was enough to get rid of the errors. I had to find other ways to include the bits I was then missing, but this was enough to get me past the issue.

**Reply by scaprile** [reply, marked as answer]
If you have a problem with includes in your environment, you have to track down your problem with your includes in your scenario. We have lots of examples that do work, and we build every day on VC for testing purposes, so unless you find there is a problem with our code, it is not.
If you read the docs, you'll see that Mongoose is told on which architecture (`MG_ARCH`) it is running, and it detects Linux, Mac and Windows.
If you see the code, you'll see that your magic words are already spoken by mongoose.h:
https://github.com/cesanta/mongoose/blob/6cf36b5771bcaec27fcc0417a23d19759c115193/src/arch_win32.h#L9-L10
so if that means anything to you, then your whatever needs that should either be after mongoose.h or you should know what you are including and why you are including it.
What mongoose includes is not an obscure set of hidden magic things, but this:
https://github.com/cesanta/mongoose/blob/6cf36b5771bcaec27fcc0417a23d19759c115193/src/arch_win32.h#L3-L134
you don't have to assume, but look at the source code
